src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,141 @@
+#include "shader_program.hpp"
+#include "utils.hpp"
+
+namespace sly::gl {
+    static constexpr char const* fallback_vertex_source{ R"(
+            #version 330 core
+            layout (location = 0) in vec3 aPos;
+
+            void main()
+            {
+                gl_Position = vec4(aPos.x, aPos.y, aPos.z, 1.0);
+            }
+        )" };
+    static constexpr char const* fallback_fragment_source{ R"(
+            #version 330 core
+            out vec4 FragColor;
+            
+            void main()
+            {
+                FragColor = vec4(255.0f, 0.0f, 255.0f, 1.0f);
+            }    
+        )" };
+
+    [[nodiscard]] constexpr std::string_view ShaderProgram::get_name_from_type(Type const type) {
+        switch (type) {
+            case Type::Vertex:
+                return "VERTEX";
+            case Type::Geometry:
+                return "GEOMETRY";
+            case Type::Fragment:
+                return "FRAGMENT";
+            default:
+                return "INVALID";
+        }
+    }
+
+    [[nodiscard]] ShaderProgram::Shader
+    ShaderProgram::compile(Type const type, std::string_view const source, bool const fallback) {
+        // compile
+
+        auto id{ glCreateShader(sly::to_underlying(type)) };
```

Comment:
```suggestion
        auto id = glCreateShader(sly::to_underlying(type));
```
Also: Can this be `const`?
```suggestion
        auto const id = glCreateShader(sly::to_underlying(type));
```

Reply:
it can not be const when i remove the recursion because i have to asign the backup shader then.
or i declare a second id. that would work too.

Reply:
If it's okay that it is not `const`, then just leave it as is. Otherwise you can use an IIFE (a lambda that you call immediately).